### PR TITLE
cv branch

### PR DESCRIPTION
**Pull request by bioblackgeorge**
train_model() changed to include also the following: gridsearch using k-fold cv, micro f1-score and hyperparameter tuning for Random Forest Classifier

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -31,10 +33,18 @@
     columns_to_drop=['building_id']
 )
 
+selected_test_values = select_features(
```
Comment (AhmedLabib98) [inline src/main.py:35]: Maybe add more comments to the code, so it is clear to everyone exactly what we are trying to achieve at each step?

File: src/encoding.py [hunk 2]
```diff
@@ -19,16 +19,16 @@ def basic_encoding(df):
...
+    # proc_df['geo_level_1_id'] = proc_df['geo_level_1_id'].astype(str)
+    # proc_df['geo_level_2_id'] = proc_df['geo_level_2_id'].astype(str)
+    # proc_df['geo_level_3_id'] = proc_df['geo_level_3_id'].astype(str)
+    # # Target Encoding the cols:
+    # geo_level_cols = proc_df['geo_level_1_id',
+    #                     'geo_level_2_id',
+    #                     'geo_level_3_id'
+    #                     ]
+    # T_encoder = TargetEncoder()
+    # proc_df[['geo_level_1_id', 'geo_level_2_id', 'geo_level_3_id']] = T_encoder.fit_transform(geo_level_cols)
```
Comment (AhmedLabib98) [inline src/encoding.py:33]: I dont really understand what the difference is, how was the code changed? It looks like you deleted and added basically the same thing

File: src/splitting.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from sklearn.model_selection import train_test_split
+
+def make_split(train_data, label_data):
+    """
+    Split values and label dataframes
+    """
+    X_train, X_valid, y_train, y_valid  = train_test_split(train_data, label_data, test_size = 0.2, random_state = 42)
+    return X_train, X_valid, y_train, y_valid
```
Comment (AhmedLabib98) [inline src/splitting.py]: This looks pretty good!

File: src/train.py [hunk 4]
```diff
@@ -1,24 +1,20 @@
...
     pipe = Pipeline([('model', model)])
 
-    # test-train split 
-    X_train, X_test, y_train, y_test = train_test_split(train_x, train_y, test_size = 0.2, random_state = 1)
-    
     # fit pipeline
-    pipe.fit(train_X, train_y)
+    pipe.fit(train_data, label_data)
 
     return pipe
```
Comment (AhmedLabib98) [inline src/train.py]: also looks good :D

File: src/train.py [hunk 5]
```diff
@@ -14,7 +20,28 @@ def train_model(train_data, label_data, model):
...
+    param_grid = { 
+        # 'model__n_estimators': [25, 50, 100, 150], 
+        # 'model__max_features': ['sqrt', 'log2', None], 
+        'model__max_depth': [3, 6, 9], 
+        # 'model__max_leaf_nodes': [3, 6, 9], 
+    } 
 
-    return pipe
\ No newline at end of file
+    # define scoring function
```
Comment (AhmedLabib98) [inline src/train.py:31]: I thought we already have a scoring function? 
Also this file is a bit big, maybe create a different python file for each function separately?

File: src/main.py [hunk 6]
```diff
@@ -28,40 +28,35 @@
...
 
 # Concat building_id and label cols into one df
-final_df = concat_pd(df_id=test_values, df_label=y_pred_test_pd)
+final_pd = concat_pd(df_id=test_values, df_label=y_pred_test_pd)
 
 # Submit
-submit(df=final_df, file_name='second_sub.csv')
-
-
+submit(df=final_pd, file_name='data/04_submissions/second_sub.csv')
```
Comment (AhmedLabib98) [inline src/main.py:61]: Great work cleaning! It looks much better now :D

**Review by AhmedLabib98 — COMMENTED** [review]
Let me know what you think about my comments :D

File: src/encoding.py (hunk 2 above)
Comment (bioblackgeorge) [reply src/encoding.py:33]: indeed, it appears confusing here, but in fact I just commented the target encoding part.

I commented it so that we remember it was not used in the current version - also since that part was not complete yet, it kinda distracted me when I was going the main.py from top to button to confirm that all individual functions are well coordinated.

File: src/main.py (hunk 1 above)
Comment (bioblackgeorge) [reply src/main.py:35]: yeah maybe we could do it - eg here # drop the building_id since it does not contain information for the training

File: src/main.py (hunk 6 above)
Comment (bioblackgeorge) [reply src/main.py:61]: heheh thank you

File: src/train.py [hunk 7]
```diff
@@ -1,24 +1,47 @@
...
     train_data: dataframe of training data
-    cols: optional list of desired columns to be considered
+    label_data: dataframe of training data
     model: a model to be trained
     """
 
     # pipeline - now basic, may add more steps here
     pipe = Pipeline([('model', model)])
 
-    # test-train split
```
Comment (AhmedLabib98) [inline src/train.py:18]: Why did you remove the train test split here?

File: src/train.py (hunk 5 above)
Comment (bioblackgeorge) [reply src/train.py:31]: the f1 function that we have in the f1_score.py is used to check the performance of our model applied on the test set. Nevertheless, cannot be used within the Gridsearchcv function that does the cross validation in the train.py (I think at least), so I created a new one. 

Sure, we could have it separately. I propose new scripts with the following:

- pipeline
- model parameters
- scoring
- gridsearch

File: src/train.py (hunk 7 above)
Comment (bioblackgeorge) [reply src/train.py:18]: originally I made it an independent function to make the train.py smaller.

Then deleted it, because cross validation is more efficient to assess and tune the model (cross validaiton basically performs splitting on its own, eg in our code now, in 5 parts - not only 2)

**Comment by bioblackgeorge** [comment]
thanks for the comments mate

File: src/main.py (hunk 1 above)
Comment (AhmedLabib98) [reply src/main.py:35]: Sure, sounds good! In general I think it would be nice that each piece of code has some comment to explain what its doing, input, output, something like that
